### ISSUE TITLE
Component: Repeating fields

### ISSUE THREAD

**Issue by Febakke** [Altinn/app-frontend-react#115]
## Description
In very simple cases our repeating groups might not be a great fit. In cases where we want the user to fill out the same field x amount of times. 

## Screenshots
**Not finished specified / Design taken from Altinn.no**
![Skjermbilde 2021-06-14 kl  16 27 55](https://user-images.githubusercontent.com/33222679/121908786-9b709080-cd2d-11eb-8bd4-950a494b926e.png)

## Considerations
- We need hidden labels for extra fields for screen readers

### Ops requirements
> Are there any requirements for monitoring? What is being built and what could go wrong? 
> Are there any requirements related to backup? 

## Acceptance criteria
- [ ] The user can add more fields
- [ ] The user can delete all fields except the first

## Specification tasks

- [ ] Development tasks are defined

## Development tasks
> Add tasks here

## Test
> Add test cases here as checkboxes that are being tested as part of the changes.

## Definition of done
Verify that this issue meets [DoD](https://confluence.brreg.no/display/T3KP/Definition+of+Done#DefinitionofDone-DoD%E2%80%93utvikling) (Only for project members) before closing.

- [ ] Documentation is updated (if relevant)
  - [ ] Technical documentation (docs.altinn.studio)
  - [ ] User documentation (altinn.github.io/docs)
- [ ] QA
- [ ] Manual test is complete (if relevant)
- [ ] Automated test is implemented (if relevant)
- [ ] All tasks in this userstory are closed (i.e. remaining tasks are moved to other user stories or marked obsolete)

**Comment by FinnurO** [comment]
@Magnusrm @olemartinorg hvordan ville dette sett ut med en repeterende gruppe med 1 felt?

**Comment by olemartinorg** [comment]
Interessant! Man kunne nok fått til noe rimelig likt via rep.gruppe med en kolonne etter denne er implementert: #579
Det store spørsmålet er om det passer med datamodellen da. I denne issuen regner jeg med man ønsker å knytte det til en datamodell som er en array med f.eks. strenger, mens en repeterende gruppe vil vel alltid være en array med objekter.

Med andre ord, datamodellen som jeg regner med ønskes brukt her:
```json
{
  "epostadresser": [
    "arild@example.com",
    "vegard@example.com"
  ]
}
```

Mens dette ville blitt resultatet med rep.grupper (med mindre vi hadde tvunget inn noe for å støtte en flatere datamodell med et felt):
```json
{
  "epostadresser": [
    { "adresse": "arild@example.com" },
    { "adresse": "vegard@example.com" }
  ]
}
```

**Comment by olemartinorg** [comment]
En ting til jeg kom på. Vi har veldig mange use-case i dag hvor vi knytter en komponent til et sted dypt inne i datamodellen, selv om det ikke finnes en sti inn dit. Som et eksempel, hvis man lager en `simpleBinding` til `A.B.C.D` vil vi putte resultatet inn i en dyp objektstruktur:

```json
{
  "A": {
    "B": {
      "C": {
        "D": "resultat"
      }
    }
  }
}
```

Så siden dette mønsteret er vel-etablert er det kanskje ikke heller forventet av en app-utvikler at vi skulle kreve at array-et i datamodellen alltid er på _nivået rett over bindingen på det repeterende elementet_. Med andre ord, vi bør ha en egen `list`-binding her, og det repeterende elementet kan ha sin egen.

Med et slikt eksempel:
```json
[
  {
    "id": "repeterende-epostadresser",
    "type": "RepeatingField",
    "dataModelBindings": {
      "list": "epostadresser"
    },
    "child": "epostadresse"
  },
  {
    "id": "epostadresse",
    "type": "Input",
    "dataModelBindings": {
      "simpleBinding": "epostadresser.adresse"
    }
  }
]
```

Hadde vi i app-frontend produsert oppføringer som:

```
epostadresser[0].adresse = "arild@example.com"
epostadresser[1].adresse = "vegard@example.com"
```

Og dermed fått det samme resultatet jeg tidligere sa var modellen som brukes i repeterende grupper. Et godt design som følger metodikken vi allerede har lagt opp til vil altså føre til at man kan benytte den strukturen i datamodellen som passer best. Denne funksjonaliteten bør uansett støtte begge varianter som jeg har nevnt over (lagring direkte til array-elementer og lagring mot objekter i array).

**Comment by Febakke** [comment]
@RonnyB71 
Yes. We have said earlier that appdevelopers should use repeating groups for this kind of scenario, but its not ideal. If this is soomething your considering we can throw together a design with the new components. The screendump in this issue is a bit outdatet.

**Comment by olemartinorg** [comment]
FYI, this is what's possible with repeating groups today:

![20230605_10h48m30s_grim](https://github.com/Altinn/app-frontend-react/assets/700139/fe0faba8-4f6e-4fa4-8066-0583e3d22d35)

**Comment by Febakke** [comment]
👏 Then I guess I dont see a reason to create a new solution for this? Maybe close this issue if there arent any limitations with this solution that hinders anyone from using it?

**Comment by olemartinorg** [comment]
I think what might be missing here is:
- A dedicated UI that makes sense for mobile. I believe, among other things, the header in the example above disappeared when the viewport shrank.
- Possibly options to make the UI less 'massive'. Right now I don't think we support `grid` options for `Group` all that well, but I'm not entirely sure either.
- The possibility to bind the repeating field directly to an array (i.e. a `list` binding instead of a `group` binding, in our terms). The example above always binds to a repeating object.
- Some nitpicks, like the ability to have more compact buttons (see #1164)

But yes, I agree that the repeating group solution may indeed solve the most pressing problems.
